### DISCUSSION TITLE
How to add this SDK as a part of Buildroot package with Gstremmer support and dependencies

### DISCUSSION THREAD

**Discussion by koratshail** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1073] [Q&A]
** Logging **
...
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vtls/mbedtls.c.o
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vtls/mesalink.c.o
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vtls/bearssl.c.o
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vquic/ngtcp2.c.o
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vquic/quiche.c.o
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vssh/libssh2.c.o
[ 16%] Building C object lib/CMakeFiles/libcurl.dir/vssh/libssh.c.o
[ 16%] Linking C shared library libcurl.so
**/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/host/opt/ext-toolchain/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/../../../../aarch64-none-linux-gnu/bin/ld: /home/ambadocker/dev/h32/output/h32_ambalink_defconfig/host/lib/libz.so: error adding symbols: file in wrong format**
collect2: error: ld returned 1 exit status
lib/CMakeFiles/libcurl.dir/build.make:2189: recipe for target 'lib/libcurl.so' failed
make[7]: *** [lib/libcurl.so] Error 1
make[7]: Leaving directory '/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/build/aws_kinesis_kvs/open-source/local/libcurl/build/src/project_libcurl-build'
CMakeFiles/Makefile2:1572: recipe for target 'lib/CMakeFiles/libcurl.dir/all' failed
make[6]: Leaving directory '/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/build/aws_kinesis_kvs/open-source/local/libcurl/build/src/project_libcurl-build'
make[6]: *** [lib/CMakeFiles/libcurl.dir/all] Error 2
Makefile:143: recipe for target 'all' failed
make[5]: Leaving directory '/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/build/aws_kinesis_kvs/open-source/local/libcurl/build/src/project_libcurl-build'
make[5]: *** [all] Error 2
CMakeFiles/project_libcurl.dir/build.make:115: recipe for target 'build/src/project_libcurl-stamp/project_libcurl-build' failed
make[4]: Leaving directory '/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/build/aws_kinesis_kvs/open-source/local/libcurl'
make[4]: *** [build/src/project_libcurl-stamp/project_libcurl-build] Error 2
CMakeFiles/Makefile2:78: recipe for target 'CMakeFiles/project_libcurl.dir/all' failed
make[3]: Leaving directory '/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/build/aws_kinesis_kvs/open-source/local/libcurl'
make[3]: *** [CMakeFiles/project_libcurl.dir/all] Error 2
Makefile:86: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/ambadocker/dev/h32/output/h32_ambalink_defconfig/build/aws_kinesis_kvs/open-source/local/libcurl'
CMake Error at dependency/libkvscproducer/kvscproducer-src/CMake/Utilities.cmake:93 (message):
  CMake step for libcurl failed: 2
Call Stack (most recent call first):
  dependency/libkvscproducer/kvscproducer-src/CMakeLists.txt:89 (build_dependency)

** Any design considerations/constraints **
I would like to add this SDK to build root system. My target architecture is AARCH64.  I am building this SDK with Dependencies, and Gstremmer support. Can you guide us how we can add this SDK into build root build system.
Following is the .mk file which I have prepared to add this SDK as build root package. Do I have to add anything else to build this SDK?
`#############################################################
#
# aws_kinesis_kvs
#
#############################################################

pkg                     = AWS_KINESIS_KVS

$(pkg)_SITE             = https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/archive/refs/tags
$(pkg)_SOURCE           = v3.3.1.tar.gz
$(pkg)_INSTALL_TARGET   = YES
$(pkg)_CONF_OPTS        = -DBUILD_GSTREAMER=ON  -DBUILD_GSTREAMER_PLUGIN=ON -DBUILD_OPENSSL_PLATFORM=linux-aarch64 -DBUILD_LOG4CPLUS_HOST=arm-linux -DBUILD_DEPENDENCIES=ON
$(pkg)_DEPENDENCIES     = openssl libcurl log4cplus gstreamer1 gst1-plugins-base

define $(pkg)_INSTALL_TARGET_CMDS
	$(INSTALL) -m 755 -D $(@D)/libgstkvssink.so $(TARGET_DIR)/usr/lib/gstreamer-1.0
	$(INSTALL) -m 755 -D $(@D)/libKinesisVideoProducer.so $(TARGET_DIR)/usr/lib
	$(INSTALL) -m 755 -D $(@D)/dependency/libkvscproducer/kvscproducer-src/libcproducer.so $(TARGET_DIR)/usr/lib
	$(INSTALL) -m 644 -D $(@D)/certs/cert.pem $(TARGET_DIR)/etc/ssl/certs/aws_root_ca.pem
endef

$(eval $(cmake-package))

`
 

** If you would not like to open an issue to discuss your solution in open-platform, please email your question to kinesis-video-support@amazon.com **

**Comment by disa6302** [comment]
@koratshail , 

We have not tried integrating with BuildRoot, so we do not have a set process in place. But, it does seem like curl build is failing on whatever system you are running this on? One of the suggestions I have is to build curl from source on the system/device and verify that build works. Alternatively you can also refer to our cross compile instructions here: https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/master/.github/workflows/ci.yml#L311

**Reply by koratshail** [reply]
@disa6302 Thanks for your reply.
I am using Ubuntu 20.04 to cross-compile this SDK. I must build this SDK with Gstremmer support.
I have tried to cross-compile this SDK standalone(without Buildroot) and without Gstremmer plugging support in your mentioned pipeline by providing CC and CXX cross-compiler flags, It successfully building. 

There are two questions,
1) How can I cross-compile this SDK with Gstremmer and Dependencies? What else must I set during compilation besides CC and CXX flags?
2) Do you have any reference to adding this SDK into the build root System with Dependencies?

Thanks in advance.

**Comment by disa6302** [comment]
1. How can I cross-compile this SDK with Gstremmer and Dependencies? What else must I set during compilation besides CC and CXX flags?

The SDK does not build GStreamer and relies on system installation. So, you would have to refer to some official documentation on GStreamer forums to understand how you could go about it. 

2. We dont have a reference. As mentioned earlier, we have not tried integrating this with build root.

**Comment by koratshail** [comment]
@disa6302 ,
What can I understand is that this SDK does not support a build system like Buildroot/Yocto. It has to be built as per the Readme.txt file given in SDK.
On top of this, SDK does not support cross-compilation if Gstremmer plugin support is on. It only relies on system installation.
Can you please confirm the above statements?

**Comment by disa6302** [comment]
@koratshail ,

We test cross compilation and building on common platforms. We do not have a specific test set up for Buildroot/Yocto.
This SDK does not compile GStreamer from source, but relies on system installation. So the platform you are cross compiling this SDK on also needs to have GStreamer cross compiled.
